src/auth/permissions/index.ts

```diff
@@ -0,0 +1,173 @@
+import * as knex from 'knex';
+import { PermissionsInitialiser, GetPermissionIds, GetRoleId } from './permissions';
+
+const getPermissionIds: GetPermissionIds = ({ client, resource, permissionLevel, access }) =>
+  client
+    .select('permission_id')
+    .table('permission')
+    .where({
+      ['permission_entity']: resource,
+      ['permission_level']: permissionLevel,
+      ['access_type']: access,
+    });
+
+const getRoleId: GetRoleId = async({ client, role }) =>
+  client
+    .select('access_role_id')
+    .table('access_role')
+    .where({ ['access_role_name']: role })
+    .then((res) => res[0].access_role_id);
+
+const permissionsInitialiser: PermissionsInitialiser = (config) => {
+  const client = knex(config);
+  return {
+    grantExisting: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
+        const permissionId = permissionIds[0].permission_id;
+
+        const insertPermission = await client('access_role_permission')
+            .returning('*')
+            .insert(
+              client.select({
+                ['access_role_id']: roleId,
+                ['permission_id']: permissionId,
+              })
+                .whereNotExists(client('access_role_permission').where({
+                  ['access_role_id']: roleId,
+                  ['permission_id']: permissionId,
+                }))
+            );
+        if (insertPermission.length === 0) {
+          throw new Error('Permission entry is already associated to this role');
```

Comment:
Might be worth sub-classing our own custom errors (list `PermissionError` or something more specific).

Reply:
By overwriting Error.name or prefixing the message with PermissionError?

Reply:
```js
class PermissionError extends Error {}
```

Reply:
Not enough for it be worth the stuff in that SO answer. I've never had to do that before though. Anyway, ignore this, it was just a thought.